include/seqan3/io/alignment_file/format_bam.hpp

```diff
@@ -0,0 +1,1061 @@
+                      "The mate object must be a std::tuple of size 3 with "
+                      "1) a std::ranges::ForwardRange with a value_type modelling seqan3::Alphabet, "
+                      "2) a std::Integral or std::optional<std::Integral>, and "
+                      "3) a std::Integral.");
+
+        static_assert(((std::ranges::ForwardRange<decltype(std::get<0>(mate))>     ||
+                        std::Integral<remove_cvref_t<decltype(std::get<0>(mate))>> ||
+                        detail::is_type_specialisation_of_v<remove_cvref_t<decltype(std::get<0>(mate))>, std::optional>) &&
+                      (std::Integral<remove_cvref_t<decltype(std::get<1>(mate))>> ||
+                       detail::is_type_specialisation_of_v<remove_cvref_t<decltype(std::get<1>(mate))>, std::optional>) &&
+                      std::Integral<remove_cvref_t<decltype(std::get<2>(mate))>>),
+                      "The mate object must be a std::tuple of size 3 with "
+                      "1) a std::ranges::ForwardRange with a value_type modelling seqan3::Alphabet, "
+                      "2) a std::Integral or std::optional<std::Integral>, and "
+                      "3) a std::Integral.");
+
+        static_assert(std::Same<remove_cvref_t<tag_dict_type>, sam_tag_dictionary>,
+                      "The tag_dict object must be of type seqan3::sam_tag_dictionary.");
+
+        if constexpr (detail::decays_to_ignore_v<header_type>)
+        {
+            throw format_error{"BAM can only be written with a header but you did not provide enough information! "
+                               "You can either construct the output file with ref_ids and ref_seqs information and "
+                               "the header will be created for you, or you can access the `header` member directly."};
+        }
+        else
+        {
+            // ---------------------------------------------------------------------
+            // logical Requirements
+            // ---------------------------------------------------------------------
+
+            if (ref_offset.has_value() && (ref_offset.value() + 1) < 0)
+                throw format_error{detail::to_string("The ref_offset object must be >= -1 but is: ", ref_offset)};
+
+            seqan3::ostreambuf_iterator stream_it{stream};
+
+            // ---------------------------------------------------------------------
+            // Writing the Header on first call
+            // ---------------------------------------------------------------------
+            if (options.sam_require_header && !written_header)
+            {
+                stream << "BAM\1";
+                std::ostringstream os;
+                write_header(os, options, header); // write header to temporary stream to query the size.
+                int32_t l_text{static_cast<int32_t>(os.str().size())};
+                std::ranges::copy_n(reinterpret_cast<char *>(&l_text), 4, stream_it); // write read id
```

Comment:
what happens here?

Reply:
The length of the reference sequence that is stored in the `ref_id_info` list of pairs at the first position is written.